bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/FindReplaceDialog.java

```diff
@@ -1895,85 +1251,92 @@ protected int getDialogBoundsStrategy() {
+		activateInFindReplacerIf(SearchOptions.CASE_SENSITIVE, fCaseCheckBox.getSelection());
+		activateInFindReplacerIf(SearchOptions.WHOLE_WORD, fWholeWordCheckBox.getSelection());
+		activateInFindReplacerIf(SearchOptions.INCREMENTAL, fIncrementalCheckBox.getSelection());
+		activateInFindReplacerIf(SearchOptions.REGEX, fIsRegExCheckBox.getSelection());
 	}
 
 	/**
 	 * Stores its current configuration in the dialog store.
 	 */
 	private void writeConfiguration() {
-		IDialogSettings s= getDialogSettings();
-
-		s.put("wrap", fWrapInit); //$NON-NLS-1$
-		s.put("casesensitive", fCaseInit); //$NON-NLS-1$
-		s.put("wholeword", fWholeWordInit); //$NON-NLS-1$
-		s.put("incremental", fIncrementalInit); //$NON-NLS-1$
-		s.put("isRegEx", fIsRegExInit); //$NON-NLS-1$
-
-		List<String> history= getFindHistory();
-		String findString= getFindString();
-		if (!findString.isEmpty())
-			history.add(0, findString);
-		writeHistory(history, s, "findhistory"); //$NON-NLS-1$
-
-		history= getReplaceHistory();
-		String replaceString= getReplaceString();
-		if (!replaceString.isEmpty())
-			history.add(0, replaceString);
-		writeHistory(history, s, "replacehistory"); //$NON-NLS-1$
+		IDialogSettings s = getDialogSettings();
+
+		s.put("wrap", findReplaceLogic.isActive(SearchOptions.WRAP)); //$NON-NLS-1$
+		s.put("casesensitive", findReplaceLogic.isActive(SearchOptions.CASE_SENSITIVE)); //$NON-NLS-1$
+		s.put("wholeword", findReplaceLogic.isActive(SearchOptions.WHOLE_WORD)); //$NON-NLS-1$
+		s.put("incremental", findReplaceLogic.isActive(SearchOptions.INCREMENTAL)); //$NON-NLS-1$
+		s.put("isRegEx", findReplaceLogic.isActive(SearchOptions.REGEX)); //$NON-NLS-1$
+
+		String findString = getFindString();
+		findHistory.add(findString);
+
+		String replaceString = getReplaceString();
+		replaceHistory.add(replaceString);
 	}
 
-	/**
-	 * Writes the given history into the given dialog store.
-	 *
-	 * @param history the history
-	 * @param settings the dialog settings
-	 * @param sectionName the section name
-	 * @since 3.2
-	 */
-	private void writeHistory(List<String> history, IDialogSettings settings, String sectionName) {
-		int itemCount= history.size();
-		Set<String> distinctItems= new HashSet<>(itemCount);
-		for (int i= 0; i < itemCount; i++) {
-			String item= history.get(i);
-			if (distinctItems.contains(item)) {
-				history.remove(i--);
-				itemCount--;
-			} else {
-				distinctItems.add(item);
+	private void activateInFindReplacerIf(SearchOptions option, boolean shouldActivate) {
+		if (shouldActivate) {
+			findReplaceLogic.activate(option);
+		} else {
+			findReplaceLogic.deactivate(option);
+		}
+	}
+
+	private void evaluateFindReplacerStatus() {
+		IStatus status = findReplaceLogic.getStatus();
+
+		String dialogMessage = status.visit(new FindReplaceLogicMessageGenerator());
+
+		fStatusLabel.setText(dialogMessage);
+
+		if (status instanceof ErrorStatus statusMessage) {
+			switch (statusMessage.getMessageCode()) {
+			case NO_MATCH:
```

Comment:
The fallthrough from `NO_MATCH` to `READONLY` does not seem to be correct:
1. `NO_MATCH` is now shown as an error rather than a message it was before.
Before: ![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/755472/8d207373-39e3-45c4-8c8f-a6aca08da873)
After: ![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/755472/f5085395-606d-42ff-8d5f-cc41d7f79355)
_Suggestion:_ Rename `ErrorStatus` to something like `FindStatus` and give it an error flag (i.e., an `isError()` method) to identify whether the status is an error or only an information.
2. `tryToBeep()` is effectively called for _every_ `ErrorStatus`. In the existing implementation, beeping is at least disabled when changing the search text in incremental search mode. This needs to be reflected in the current implementation: https://github.com/eclipse-platform/eclipse.platform.ui/blob/f89725841c5b9c780c628bf80c66ef2ad716158c/bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/FindReplaceDialog.java#L155
https://github.com/eclipse-platform/eclipse.platform.ui/blob/f89725841c5b9c780c628bf80c66ef2ad716158c/bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/FindReplaceDialog.java#L1534
In fact, no status evaluation is performed when modifying the text for incremental search, as `FindModifyListener#modifyText` only performs a search operation but no status evaluation like all the button handlers do. This needs to be added and then `evaluateFindReplacerStatus` should be parameterized with a `beep` parameter to reflect the existing beeing enablements.

Reply:
I have tweaked this behaviour and I believe that I match the current behaviour correctly.
In particular, now I don't display "No Match found" and "wrapped" as errors and I don't beep on "wrapped search". I also stop beeping on incremental search but **do** display the messages as if the search had occurred normally.

https://github.com/eclipse-platform/eclipse.platform.ui/commit/555a45e5a86580594a0b35108879fa33b98de879